src/app/donation-start/donation-start-form/donation-start-form.component.ts

```diff
@@ -689,6 +697,64 @@ export class DonationStartFormComponent implements AfterContentChecked, AfterCon
     }
   }
 
+  private runFinalPreSubmitUpdate() {
+    // Even if next guard fails, we want to prevent attempting to pay.
+    this.runningFinalPreSubmitUpdate = true;
+
+    let requiredDonationProperties = [this.donation, this.campaign, this.campaign.charity.id, this.psp];
+    // Return if any of the required properties are missing.
+    requiredDonationProperties.forEach((value, index) => {
+      if (!value) {
+        const errorCodeDetail = `[code A${index}]`; // Data at `requiredDonationProperties` index `index` absent.
+        const errorMessage = `Missing donation information – please refresh and try again, or email hello@biggive.org quoting ${errorCodeDetail} if this problem persists`;
+
+        this.matomoTracker.trackEvent(
+          'donate_error',
+          'submit_missing_donation_basics',
+          `Donation not set or form invalid ${errorCodeDetail}`,
+        );
+        this.toast.showError(errorMessage);
+        this.donationUpdateError = true;
+        this.stripeError = errorMessage;
+        this.stripeResponseErrorCode = undefined;
+      }
+    });
+
+    // First part of clause should be redundant but TS doesn't track enough to know that(?)
+    if (!this.donation || this.donationUpdateError) {
+      return;
+    }
+
+    if (this.paymentGroup.value.billingPostcode) {
+      this.donation.billingPostalAddress = this.paymentGroup.value.billingPostcode;
+      this.donation.countryCode = this.paymentGroup.value.billingCountry;
+      this.donationService.updateLocalDonation(this.donation);
+    }
+
+    this.donationService.update(this.donation)
+      .pipe(retry({ count: 3, delay: getDelay() }))
```

Comment:
Keeping a retry but replacing the deprecated RxJS operator use, and slightly simplifying our delay timing logic for legibility